### PR TITLE
Title of newly uploaded package should not match existing package registration id

### LINKED ISSUE #2849: Title of newly uploaded package should not match existing package registration id
Several checks should be done on upload/edit:

* Title of newly uploaded package should not match existing package registration id
* New package registration id should not match an existing package title

### PR DESCRIPTION
**Pull request by maartenba**
Fixes #2849

On upload/edit package:

* Title of newly uploaded package should not match existing package registration id
* New package registration id should not match an existing package title

/cc @yishaigalatzer @joyhui @joelverhagen

### CODE REVIEW

**Comment by xavierdecoster** [comment]
:shipit:

File: src/NuGetGallery/Services/PackageNamingConflictValidator.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            _packageRegistrationRepository = packageRegistrationRepository;
+            _packageRepository = packageRepository;
+        }
+
+        public bool TitleConflictsWithExistingRegistrationId(string registrationId, string packageTitle)
+        {
+            if (string.IsNullOrEmpty(registrationId))
+            {
+                throw new ArgumentNullException("registrationId");
```
Comment (maartenba) [inline src/NuGetGallery/Services/PackageNamingConflictValidator.cs]: Will be updated to `nameof(registrationId)` once https://github.com/NuGet/NuGetGallery/pull/2846 gets enough :shipit:

File: src/NuGetGallery/Services/PackageNamingConflictValidator.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            return false;
+        }
+
+        public bool IdConflictsWithExistingPackageTitle(string registrationId)
+        {
+            if (string.IsNullOrEmpty(registrationId))
+            {
+                throw new ArgumentNullException("registrationId");
```
Comment (maartenba) [inline src/NuGetGallery/Services/PackageNamingConflictValidator.cs]: Will be updated to `nameof(registrationId)` once https://github.com/NuGet/NuGetGallery/pull/2846 gets enough :shipit:

**Comment by maartenba** [comment]
Please review, @joyhui @yishaigalatzer @johnataylor @xavierdecoster

File: src/NuGetGallery/Services/PackageNamingConflictValidator.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            if (!string.IsNullOrEmpty(packageTitle))
+            {
+                var cleanedTitle = packageTitle.Trim().ToLowerInvariant();
+                if (!string.IsNullOrEmpty(cleanedTitle))
+                {
+                    var packageRegistration = _packageRegistrationRepository.GetAll()
+                        .SingleOrDefault(pr => pr.Id.ToLower() == cleanedTitle);
+
+                    if (packageRegistration != null 
+                        && !String.Equals(packageRegistration.Id, registrationId, StringComparison.OrdinalIgnoreCase))
```
Comment (xavierdecoster) [inline src/NuGetGallery/Services/PackageNamingConflictValidator.cs]: nit: lowercase `string` in String.Equals

**Comment by xavierdecoster** [comment]
:shipit:

**Comment by joyhui** [comment]
:shipit:

**Pull request merged**